File: algorithms/07_typescript/sorting/insertion_sort.ts
export function insertionSort(arr: number[]): number[] {
  const a = arr.slice();
  for (let i = 1; i < a.length; i++) {
    const key = a[i];
    let j = i - 1;
    while (j >= 0 && a[j] > key) { a[j + 1] = a[j]; j--; }
    a[j + 1] = key;
  }
  return a;
}

const sorted = insertionSort([5, 2, 4, 6, 1, 3]);
if (sorted.join(",") !== "1,2,3,4,5,6") throw new Error("Insertion sort failed");
console.log("[TypeScript Insertion Sort] Sorted array verified");

File: algorithms/08_r/sorting/insertion_sort.R
insertion_sort <- function(arr) {
  n <- length(arr)
  if (n <= 1) return(arr)
  for (i in 2:n) {
    key <- arr[i]
    j <- i - 1
    while (j >= 1 && arr[j] > key) {
      arr[j + 1] <- arr[j]
      j <- j - 1
    }
    arr[j + 1] <- key
  }
  arr
}

# Example
stopifnot(identical(insertion_sort(c(5, 2, 4, 6, 1, 3)), c(1, 2, 3, 4, 5, 6)))
print("[R insertion_sort] ... verified")
